### PR TITLE
Replace Mobility table with log equations (with rook mg exception).

### PR DESCRIPTION
**Pull request by protonspring**
The equation for each piece was carefully tested with each passing -3,1 STC/LTC as well as in combination.  The result is a version that appears to be slightly better than master, but does not pass 0,4.

Log equations for mobility makes sense because lower mobility values are much more sensitive than higher mobility values (mobility of 1 us MUCH worse than mobility of 3 and mobility of 15 is not much different from mobility of 14).  Note that the original table values were very logarithmic and these equations very closely resemble the original curves.

The only exception is rooks mg which was linear except the value of mobility=0, the exception is in the code.  Please see the comments which include more detailed analysis, etc.

It is hoped that less values will be easier to tune/manage than an entire array of values.
Although there are more code lines, there is less code due to comments and white space.

Furthermore, there is reason to believe that the offsets could eventually be simplified away using the psqt table although my original attempts failed.

Should i link in all of the passing tests for each piece?

LTC - (combinaton of all equations)
LLR: 2.96 (-2.94,2.94) [-3.00,1.00]
Total: 28279 W: 4731 L: 4622 D: 18926 
http://tests.stockfishchess.org/tests/view/5bcb64730ebc592439f89004

Please let me know how I should proceed.

### CODE REVIEW

**Comment by NKONSTANTAKIS** [comment]
I agree with @Rocky640 , @protonspring and @31m059. The scattered values of the arrays are no match for the fluid values of the mathematical formulas. Concentrated information is more simple and powerful for the processing, maybe the look complex to us but its not like we are going to hand tune anything at this level, all comes from SPSA etc. And the formulas are easier tunable with far less values as mentioned.

File: src/evaluate.cpp [hunk 1]
```diff
@@ -852,9 +852,8 @@ namespace Eval {
            + Eval::Tempo;
   }
 
-/// Eval::init() initializes some tables needed by evaluation. Instead of using
-/// hard-coded tables, when makes sense, we prefer to calculate them with a formula
-/// to reduce independent parameters and to allow easier tuning and better insight.
+/// Eval::init() initializes some tables needed by evaluation. A formula reduces
+/// indepednent parameters and allows easier tuning.
```
Comment (jerrydonaldwatson) [inline src/evaluate.cpp]: Note a misspelling: "independent" not "indepednent"

File: src/evaluate.cpp [hunk 2]
```diff
@@ -865,6 +852,19 @@ namespace {
            + Eval::Tempo;
   }
 
+/// Eval::init() initializes some tables needed by evaluation. A formula reduces
+/// independent parameters and allows easier tuning.
+void init() {
+
+  for (int m = 0; m < 32; ++m)
+  {
+    MobilityBonus[ QUEEN-2][m] = make_score(52-120*(1-log10(m+2.0)),  48-270*(1-log10(m+5.0)));
```
Comment (vondele) [inline src/evaluate.cpp]: is there a particular reason to write '52-120*(1-log10(m+2.0))' ? Equivalent would be '-64+120*log10(m+2.0)'

Comment (protonspring) [reply src/evaluate.cpp]: That's just the first form that came into my head when I was trying to replicate the curves.  Your version (although not exact) is certainly formed better. . I will try to duplicate that form. . .

**Comment by vondele** [comment]
looks good to me.

**Comment by protonspring** [comment]
Since these are now equations, I guess they could just be calculated while building the psq per square array.  This would remove the BONUS array altogether.  Should I make that change?

**Comment by Rocky640** [comment]
I don't get your idea. We still need a MobilityBonus (by piece type and mobility) array in evaluate.cpp.
How does it relate to the Bonus array (by piece type and square) of psqt.cpp ?

**Comment by protonspring** [comment]
@Rocky640 sry . . brain fart. . . I got mixed up between Mobility and psqt Bonus.  disregard.

**Comment by Rocky640** [comment]
I suggest that you move the values used by your formula to a static int A[] = {...};
and also define a static double B[] ={...}; (for the log or the slope offsets). This will facilitate further tuning attempts. Also add some spacing around each arithmetic operator.

**Comment by protonspring** [comment]
My first time using a lambda object!  It should be compatible with everything since C++11.  Hopefully, this is OK.  It does make the code a bit cleaner.

**Comment by protonspring** [comment]
Note:  This passed, which may affect this PR.

http://tests.stockfishchess.org/tests/view/5bd3188f0ebc595e0ae1e296

**Comment by Vizvezdenec** [comment]
As per usual I guess that elo gain > simplification.
Also note what I've written in PR to this test - I have a feeling that there is some more elo to be squeezed from this parameters so simplifying them into formula is not the way to go for now, imho.

**Comment by protonspring** [comment]
Possibly. . Perhaps some equation tuning is warranted.  It would be
interesting to see how it compared.

**Comment by Vizvezdenec** [comment]
Sure.
Although SPSA is usually better with tuning large arrays. So probable way to go will be smth like this
1) retune knight parameters at high TC;
2) check for elo gained (if any);
3) then try to simplify it (if it's the direction we decide to go with this stuff).
Although we need someone to do first step because I don't know how to use SPSA :)

**Comment by mcostalba** [comment]
This is a case where the simplification is really in the eye of the beholder. Personally, I fully second Stéphane judgment: this is not a simplification, so it should be tested as an ELO-gainer: SPRT[0, 5]

You are replacing a bunch of numbers with another bunch of numbers (although less), with added formulas and loops. The end result is both more code and more complex logic.

Sorry but this is **not** a simplification to me.

**Comment by vondele** [comment]
A quote from pawns.cpp:
```
/// Pawns::init() initializes some tables needed by evaluation. Instead of using
/// hard-coded tables, when makes sense, we prefer to calculate them with a formula
/// to reduce independent parameters and to allow easier tuning and better insight.
``` 
why is this any different? At least the proposed version (I like https://github.com/official-stockfish/Stockfish/pull/1784/commits/8a230ea99b36f896391003a2c433988ece0145c8) makes clear that we have a simple logarithmic dependence (which insight and makes sense) and reduces the number of parameters we have significantly.

**Comment by mcostalba** [comment]
// we prefer to calculate them with a formula
/// to reduce independent parameters and to allow easier tuning and better insight.

This is more wishful thinking to me than a proven fact. Namely it is not clear that tuning with less parameters leads easily to improved ELO. Actually, in the past, admittedly to my great surprise, I experienced many cases in which our tuning algorithm worked better with many params than with few of them. But because tuning is hard, I have not a final opinion on it, just that the above sentence is simply an assumption, not a proven fact. And because instead it is put as a proven fact, maybe it is misleading too.

**Comment by vdbergh** [comment]
@mcostalba 

You probably recall that when someone first attempted to tune multiple parameters you called it an 

"epic resource draining".

Will you now apologize for making such an unfounded statement?

**Comment by xoto10** [comment]
But do we have a logarithmic dependence? I'm not comfortable that a log
curve is the right one and am concerned that it distorts the results too
much.
For example, i can believe the difference between a knight having a
mobility of 0 or 1 is small, and also the difference between 7 and 8 is
small, but the values inbetween have larger differences. So that might be
some kind of logistic equation.

**Comment by protonspring** [comment]
It is VERY clear that all of the mobility values are log based.  Plot them
and see for yourself.

**Comment by NKONSTANTAKIS** [comment]
I have only seen the graphs for Queen, but I believe you of course that all the current ones are log based. The mild concern was about flexibility for the future in case something could be better by not being log based. What about the new passed knight values, as a check of this, are they also logarithmic?

**Comment by protonspring** [comment]
Can you not just plot the values? Copy them in a spreadsheet and check them
out.

**Comment by NKONSTANTAKIS** [comment]
First time I do this, thanks for motivation. Blue is old values, Red is new(Viz), Black is log scale of new.

knight mg
knight eg

**Comment by protonspring** [comment]
Nicely done!  If you do the others you will see that all of them are very
logarithmic.

**Comment by NKONSTANTAKIS** [comment]
Now that I got the hang of it
bishop mg
bishop eg

**Comment by protonspring** [comment]
It is starting to look like, the log equation for knights is not as good as the values.  Thus, due to the uniqueness of the knight, this equation does not work for all pieces.

I can make an equation for knight, but it's more complicated and I'd rather have the values.

I also would not want some pieces to have numeric tables, and others to have equations, so I'd be inclined to *sadly* close this PR without commit.

Agree/disagree?

**Comment by 31m059** [comment]
@protonspring I agree.

**Comment by Vizvezdenec** [comment]
I think that replacing values with equations also has another drawback - for some reason SPSA tuning produces better results with big arrays rather than with small set of values. So replacing mobility tables with equations will counter-intuitively reduce tunability (!).

**Comment by 31m059** [comment]
@Vizvezdenec I'm not entirely sure of this--there's a confound we're missing.  In general, we tend to tune large arrays at LTC with very long tuning sessions; small sets of parameters are often tuned with shorter TC and even more often with fewer games.

If we tune an equation and an array at the same TC for the same number of games, I'm not sure the array will perform better.

**Pull request closed without merging**